[View] Keep view's aspect ratio based on new prop

paramaggarwal opened the pull request:
Continues from #858.

Current behaviour:
1. When a static image is configured with `<Image source=require("image_name.jpg") />`, the packager adds the `width` and `height` into the `source` prop. Hence there is no need of setting dimensions in `style`.
2. But for remote images, this `width` and `height` is forcefully set on the `style` prop, which means if one uses only `alignSelf: 'stretch'` to layout the image, the `height` would not change in the same ratio as the `width` inherited from the parent.

This pull request fixes this. I believe the basic problem that this attempts to solve is very genuine. It includes a `measure` function on NetworkImageView to maintain aspect ratio of image when dimensions are known beforehand.

Now just like automatic sizing and layout of text, the same would work for images that actually take the entire width of the device, and the height would be configured automatically based on the calculated aspect ratio of the image:
```jsx
<View style={{flex: 1, paddingTop: 20}}>
  <Image
    style={{margin: 5}}
    source={{
      uri: "http://placehold.it/800x100",
      height: 100,
      width: 800
    }}
  />
  <Image
    style={{margin: 5}}
    source={{
      uri: "http://placehold.it/600x300",
      height: 300,
      width: 600
    }}
  />
  <Image
    style={{margin: 5}}
    source={{
      uri: "http://placehold.it/400x400",
      height: 400,
      width: 400
    }}
  />
</View>
```
![Screenshots comparing the output of above code on devices with different widths.](http://i.imgur.com/XwvXTGY.png)

Open to suggestions on how to improve this pull request.

brentvatne commented:
@paramaggarwal - this is a really interesting idea! cc @vjeux

ide commented:
I think we want this in core since it achieves parity between static and network images and it's a fundamental that applies to all images. To be clear, @paramaggarwal's proposal maintains the current invariant of needing to specify the image size up front (before we've downloaded it from the network). The sole difference is that it's the intrinsic image size, not the displayed size (in the style prop) that needs to be specified. This provides a very consistent API for users that I could explain to people:

 - You should specify an image's intrinsic dimensions in the `<Image source>` prop
 - The intrinsic dimensions, which tell us the intrinsic aspect ratio, will be used to help the layout engine calculate the image's displayed dimensions (e.g. the flex example in this discussion)
 - If you don't specify intrinsic dimensions then you must specify the image's displayed dimensions via width and height in the `style` prop.
 - The packager automatically specifies the intrinsic dimensions when it replaces `require('image!icon.png')`
 - Missing intrinsic dimensions default to zero

@paramaggarwal: might want to rename original width/height to intrinsic width/height to match w3c docs.

brentvatne commented:
Great points @ide - what are your thoughts on this @vjeux? I know you've looked into this kind of thing previously.

paramaggarwal commented:
Thanks @brentvatne and @ide. I have renamed originalHeight/originalWidth to intrinsicHeight/intrinsicWidth. I have also updated the calculated dimensions to be `0x0` when intrinsic dimensions are not provided instead of the previous assumption of a default aspectRatio of `4:3`.

For the benefit of others, here is the flow of how the final dimensions of a network image are calculated in order of priority:
1. Absolute `height` and `width` specified in `style`.
2. Flex calculated height and width coming from `style`.
3. Intrinsic height and width used from the dimensions passed to `source`.

In all of these scenarios, if it is not able to calculate the height from the given info, it will attempt to calculate by using the aspect ratio computed from the dimensions passed to `source`.

Travis test is failing because of this failure on `TEST_TYPE=build_website` over [here](https://travis-ci.org/facebook/react-native/jobs/65834149#L5298), and not in the `obj-c`, `js` tests.
 ```
{ [Error: socket hang up] code: 'ECONNRESET' }
```

sahrens commented on Libraries/Image/Image.ios.js:
```diff
@@ -143,6 +139,14 @@ var Image = React.createClass({
+    var style = flattenStyle([{width, height}, styles.base, this.props.style]);
+    if (isNetwork) {
+      // NetworkImageView uses a measure method instead to set original width and height
+      style = flattenStyle([styles.base, this.props.style]);
```
you're throwing away the work done in flattenStyle above - it can be quite expensive, so only do one or the other.

paramaggarwal replied:
Good one, I fixed this.

sahrens commented:
This is cool, but we have some changes planned with the source width/height that may need to coordinate here (although I think this is a better solution to the problems we've been seeing than we were planning).  Assigning to @frantic who has been working on that image stuff.

sahrens commented:
Also, can you rebase?  It might also give a passing test run.

paramaggarwal commented:
@sahrens Squashed commits into one, rebased to fix merge conflicts, removed redundant `flattenStyle` and tests now passing!

paramaggarwal commented:
@frantic please share your thoughts.

Also, I noticed while using this, that setting a `measure` function on css_layout will not cause it to look for measurements based on child elements. So if you have an image with children, and you would like the image to get its dimension based on children, then this is a no-go.

Currently I do this:
```
- (void)fillCSSNode:(css_node_t *)node
{
  [super fillCSSNode:node];
  if (_intrinsicHeight && _intrinsicWidth) {
    node->measure = RCTMeasure;
  }
}
```

So this works well. It works like an opt-in switch. Passing in `source.width` and `source.height` will mean that you want the image to size itself based on its aspect ratio than its children. The problem is that if an image element was first configured without this, and then later on gets the  `source.width` and `source.height` parameters, this aspect ratio behaviour would not get turned on. Thoughts?

sahrens commented:
> The problem is that if an image element was first configured without this, and then later on gets the source.width and source.height parameters, this aspect ratio behaviour would not get turned on.

In general, it's really important that the order of property changes not matter, just the final state - can you fix that?

paramaggarwal commented:
@sahrens, I thought of a way to do it conditionally. Now only if an `intrinsicHeight` or `intrinsicWidth` are passed in will it attempt to measure layout dynamically.

@frantic @vjeux please do share your thoughts.

vjeux commented:
Sorry for the super long delay in my response.

Can we polyfill this on web somehow? Two ways would be padding-bottom: (aspectRatio * 100)% or using <img> with a data uri that's a transparent image of the intrisic size.

If we can't then I'm going to push back on this. Otherwise this seems like a fine change. Very much opt-in.

For your measure question, this is indeed an issue, we probably need to change the caching logic to let a component defined function also tell if it should invalidate caching or not (based on intrisic width)

vjeux commented:
Btw image themselves may not be the only one that require an aspect ratio. If you want to do a responsive square grid, you need to maintain 1:1 aspect ratio on a View. Out of the scope of this pull request but would be nice to unify the concept of aspect ratio constraint

paramaggarwal commented:
Isn't this the default behaviour on the web already? When an `<img>` is used on the web, the browser fetches the image, calculates the width and height and maintains the aspect ratio of the image automatically. See [this JSBin](http://jsbin.com/jalewegugo/edit?html,css,output), try resizing the output panel and note that when the width of the image is changing, the height is also changing. Hence the aspect ratio of the image is automatically maintained by the browser.

Now in React Native, we do not load images and calculate dimensions, we expect the dimensions to be passed before-hand. But that causes us to lose this capability where changing the width of the image, changes the height accordingly.

This PR, helps us give the original `intrinsicHeight` and `intrinsicWidth` of the image (which a browser finds out after actually loading the image, while we provide beforehand) - and based on this, using the `measure` method of css-layout, we are able to maintain the aspect ratio of the image when it's width changes.

In much this way, the `src` tag on the web, maps to the `source` tag in React Native. Only that, in addition to the `uri`, we also need to give the `width` and `height` of the image - completely describing the remote resource.

```
<Image source={{
  uri: 'https://upload.wikimedia.org/wikipedia/commons/thumb/5/57/React.js_logo.svg/600px-React.js_logo.svg.png',
  width: 600,
  height: 600
}} />
```

frantic commented:
Hey @paramaggarwal, great work!

I'm wondering if we could use the same logic for both static and network images? Or am I missing something?

For more context about the image work we are doing internally. The idea is to deprecate `require('image!...')` usage in favor of relative require styles (`require('./logo.png')`). There are lots of benefits to this system: no need to manually add images in Xcode, no need to recompile, images live alongside the code, etc. You can follow it here #1043

Implementation wise we will be serving the images over network in development mode and will bundle them with the app when submitting to AppStore, so the same image will be rendered using NetworkImage and StaticImage depending on environment. That means they have to behave the same way to avoid surprises.

As part of that, we made packager add width and height to image metadata, so the developer won't need to duplicate information that the system can figure out. The only problem with this approach is when the developer wants to use something like flex to size the image - width and height attributes we set get priority. One solution could be flattening style and checking if flex or alignSelf are set, and only if not - applying known width and height.

ide commented:
> I'm wondering if we could use the same logic for both static and network images?

Yes, this should absolutely happen.

- For truly remote images (`http://example.com/logo.png`) the programmer has to specify `{width, height, uri}` in the image source
- For images that are locally available but are served remotely by the packager (`./logo.png`), the packager should read the intrinsic size of the image and replace `require('./logo.png')` with `{width, height, uri}`
- For images that are locally available and are statically bundled, the packager should read the intrinsic size of the image and replace `require('./logo.png')` with `{width, height, uri, isStatic}`

Then, both the native implementations for static and remote images should take the provided intrinsic dimensions into account when computing layout.

> The only problem with this approach is when the developer wants to use something like flex to size the image - width and height attributes we set get priority.

The width and height properties of the image source should not get priority. They are intrinsic dimensions, not the rendered dimensions.

Just like the web, they should be the defaults that the layout engine uses as a last resort. For example, if the programmer uses flex and we know the rendered width of the image should be 375px wide but don't know its rendered height, then the layout engine should calculate `rendered_height = rendered_width * intrinsic_height / intrinsic_width`. If the programmer specifies `style={{width: 200}}` but leaves off the height, then the layout engine should perform a similar calculation using the intrinsic aspect ratio to compute the rendered height. If the programmers doesn't specify any rendered dimensions (no flex, no width nor height in the `style`), then the image falls back to its intrinsic dimensions.

paramaggarwal commented:
Thanks @frantic, the behaviour of static and remote images is now exactly the same. I've published an example to https://github.com/paramaggarwal/flex-image-demo to demonstrate this.

Thanks @ide, the behaviour you describe is exactly what happens here, please try out the example I have published. Try experimenting with `index.os.js`.

(Travis refusing to run the test as the build on master is failing at the moment.)

paramaggarwal commented:
I am trying the test cases described by @ide above.

> 1. If the programmer uses flex and we know the rendered width of the image should be 375px wide but don't know its rendered height, then the layout engine should calculate rendered_height = rendered_width * intrinsic_height / intrinsic_width.

This works, but without having to explicitly use `flex`. More details below.

> 2. If the programmer specifies style={{width: 200}} but leaves off the height, then the layout engine should perform a similar calculation using the intrinsic aspect ratio to compute the rendered height.

This works as expected.

> 3. If the programmers doesn't specify any rendered dimensions (no flex, no width nor height in the style), then the image falls back to its intrinsic dimensions.

This does not work as expected. When no `style` dimensions are specified, it is picking `width` as the width of the parent.

I attempted to fix this by trying these modifications in my code:
1. If the width of the image is smaller than the width of parent, then return the width of the image as is, with the height calculated in proportion.
But this still didn't work because css-layout uses only the height from the returned dimensions and not the width. It continues to use the width of the parent itself. The relevant snippet of code is [here](https://github.com/facebook/css-layout/blob/master/src/Layout.c#L575-L587).

2. If there was a way to check if `flex` is being used on the node, then return intrinsic dimensions as is.
Doing this is hard from outside of css-layout, as detecting whether `flex` is being used can only be determined from inside the css-layout code. This is because `flex` could be cascading down the children.

@vjeux please help with your thoughts here. Should this kind of layout be a method on css-layout instead? For example right now, we specify a `measure` function for css-layout, but instead if we could pass the intrinsic dimensions of the node to css-layout itself? That way this could be applied to any node, not just images. Please do share thoughts.

frantic commented:
@ide great writeup! @paramaggarwal I'm happy with your approach, will let @vjeux stamp this.

paramaggarwal commented:
The `measure` method capability exposed by `css-layout` was originally intended only for text. This allows any node to define a function that is passed a width, and it is able to return a width and height at runtime. Hence `RCTText` nodes use this, and don't need explicit width height styles. Text simply takes up the amount of height that it needs. This is exactly how it is on the browser too.

Now to use this capability for images, and support for browsers treat images, we hit into the two following limitations of the `measure` method.

1. It uses the width of the parent node, irrespective of flex being active or not.
2. It uses the width of the parent node, irrespective of width returned from the `measure` method.

That is, it is very adamant about the width being equal to the parent node, unless explicitly defined in `style`. It does not even expect an `alignSelf: 'stretch'` on the node or `alignItems: 'stretch'` on the parent. This makes perfect sense for the original use case of a text node, but not for an image node.

To be able to do layout for image nodes, either we modify the above two behaviours. But I realised that the math we do for image layout is pretty simple, and does not depend on platform specific methods (how in `measure` we need to use platform specific APIs for text sizing.)

Hence, instead of using defining a `measure` method, if one could set `intrinsicDimensions` of a css-node, then `css-layout` could automatically do the math, which would be as follows (from @ide's comment):

1. If `flex` active, calculate child `width` based on parent `width` and child `height` in proportion.
2. If `width` dimension specified in `style`, then calculate `height` using `intrinsicDimensions`.
3. If no `style`, use `intrinsicDimensions` as is.

paramaggarwal commented:
I just successfully tried out a standalone `<AspectRatio />` component. It looks like this, and does not require any changes to the <Image /> module. It is also completely opt-in and for generic use on any component whatsoever. The way it works is that it takes the input dimensions and uses them to size itself into the constraints of the parent view. Hence always maintaining the specified aspect ratio.

```
<AspectRatio dimensions={{width: 980, height: 480}}>
  <Image style={{flex: 1}} source={{uri: "http://example.com/image.jpg"}} />
</AspectRatio>
```

Unfortunately, I cannot publish as it needs this change from css-layout imported into React Native: https://github.com/facebook/css-layout/pull/84.

vjeux commented:
@paramaggarwal: omg this is awesome! Having a standalone component is the best possible outcome, I can't wait to see your implementation

Can you submit a pull request that just add this css-layout change to react-native repo? I'll happily stamp and import it.

paramaggarwal commented:
Thanks @vjeux. I have submitted #1776, updated this PR and uploaded a demo here: https://github.com/paramaggarwal/aspect-ratio-demo.

It looks like this for an `Image`:
```
<Image
  aspectRatioBox={{
    width: 612,
    height: 456,
  }}
  source={{
    uri: "http://example.com/image.jpg",
  }}
/>
```

and a `View` that should always be a square, and then flex layout everything inside it:
```
<View aspectRatioBox={{
    width: 1,
    height: 1,
  }}>
    <View style={{flexDirection: 'row', flex: 1}}>
      <View style={{flex: 1, backgroundColor: 'red'}} />
      <View style={{flex: 2, backgroundColor: 'green'}} />
    </View>
    <View style={{flexDirection: 'row', flex: 1}}>
      <View style={{flex: 3, backgroundColor: 'blue'}} />
      <View style={{flex: 2, backgroundColor: 'yellow'}} />
    </View>
  </View>
```

I thought about it, and decided to make this an opt-in `aspectRatioBox` prop on any generic node because it seemed like a valid use-case to actually be able to use this anywhere at all. Please do share your thoughts in the naming of the `aspectRatioBox` prop and the place where this code should exist.

vjeux commented:
The fact that the argument is a vector is going to be pretty hard to convey. aspectRatio is scalar and giving it a vector is tricky. Also, one very common use case is to want a square aspect ratio which is awkward with your API.

One thing we can do is to name it aspectRatio and be able to take either a number or a vector. This way both use cases are properly covered and you can use the one that's most appropriate.

vjeux commented on Libraries/Components/View/View.js:
```diff
@@ -177,6 +177,18 @@ var View = React.createClass({
+     * to wrap images and other components, where you would like the aspect
+     * ratio to be maintained.
+     */
+    aspectRatioBox: PropTypes.shape({
```
you also need to add this to <Image>

vjeux commented on React/Views/RCTShadowView.m:
```diff
@@ -548,6 +570,19 @@ - (type)getProp                                             \
+- (void)setAspectRatioBox:(CGSize)aspectRatioBox {
+    _aspectRatioBox = aspectRatioBox;
+    if (aspectRatioBox.width && aspectRatioBox.height) {
+        self.cssNode->measure = RCTAspectRatioMeasure;
```
nit: indent 2 spaces

vjeux commented on React/Views/RCTShadowView.m:
```diff
@@ -86,6 +86,28 @@ static void RCTProcessMetaProps(const float metaProps[META_PROP_COUNT], float st
+ */
+static css_dim_t RCTAspectRatioMeasure(void *context, float width)
+{
+    RCTShadowView *shadowView = (__bridge RCTShadowView *)context;
```
nit: 2 spaces

vjeux commented on React/Views/RCTShadowView.m:
```diff
@@ -86,6 +86,28 @@ static void RCTProcessMetaProps(const float metaProps[META_PROP_COUNT], float st
+static css_dim_t RCTAspectRatioMeasure(void *context, float width)
+{
+    RCTShadowView *shadowView = (__bridge RCTShadowView *)context;
+    if (isnan(width)) width = shadowView.aspectRatioBox.width;
```
nit: please always put braces and the statement on the next line

vjeux commented on React/Views/RCTShadowView.m:
```diff
@@ -86,6 +86,28 @@ static void RCTProcessMetaProps(const float metaProps[META_PROP_COUNT], float st
+    }
+    
+    css_dim_t result;
+    result.dimensions[CSS_WIDTH] = ceilf(width);
```
Please don't round here. We only want to round exactly once when we apply it to the view, otherwise we're losing precision all along the way.

And, if we were to round here, we would need to ask many questions like
- why ceil instead of floor or round?
- the pixel density needs to be taken into account. `1` unit is equivalent to `2` physical pixels on iphone 5 for example. If you were to do this, you would lose half your precision

ide commented:
Regarding the prop name I'd vouch either for `intrinsicSize={size}` or `aspectRatio={number}`. I can see scenarios in which the intrinsic size is useful (images, also 3d views if RN ever wants to do that) and scenarios in which a component author only wants to provide an aspect ratio (vector graphics).

I think the answer is that we might want both (and it is a warning if they don't match). http://dev.w3.org/csswg/css-images-3/#intrinsic-dimensions says 
> Raster images are an example of an object with all three intrinsic dimensions. SVG images designed to scale might have only an intrinsic aspect ratio; SVG images can also be created with only an intrinsic width or height.

paramaggarwal commented:
@vjeux thanks for the in-depth review - I have incorporated all your feedback.
@ide @vjeux `aspectRatio` with type-checking sounds perfect, I've implemented this:

```
aspectRatio: React.PropTypes.oneOfType([
  React.PropTypes.number,
  React.PropTypes.shape({
      width: React.PropTypes.number,
      height: React.PropTypes.number,
    }),
]),
```

I've also updated the demo here: https://github.com/paramaggarwal/aspect-ratio-demo.

ryangomba commented:
I like this a lot and have found it useful myself. However, you're missing RCTShadowView's setAspectRatio implementation in this diff :)

paramaggarwal commented:
@ryangomba thank you so much! Missed it while copying over changes.

paramaggarwal commented:
Please review!

brentvatne commented:
@paramaggarwal - I think he is on vacation still but I think he should be available next week :)

brentvatne commented:
ping @vjeux :heart_eyes:

dubert commented:
:+1:

paramaggarwal commented:
@dubert Nope, not yet. But might be useful data to have on let's say the `onLoad` method of an `<Image>` tag. This is not implemented yet.

paramaggarwal commented:
ping @vjeux @ide @sahrens. Please share your thoughts!

ide commented on React/Views/RCTViewManager.m:
```diff
@@ -275,4 +275,14 @@ - (RCTViewEventHandler)eventHandlerWithName:(NSString *)eventName json:(id)json
 
 RCT_REMAP_SHADOW_PROPERTY(onLayout, hasOnLayout, BOOL)
 
+RCT_CUSTOM_SHADOW_PROPERTY(aspectRatio, CGSize, RCTShadowView) {
```
I believe you should handle the case where `json` is nil with `defaultView.aspectRatio`.

paramaggarwal replied:
fixed

ide commented on React/Views/RCTShadowView.m:
```diff
@@ -86,11 +86,46 @@ static void RCTProcessMetaProps(const float metaProps[META_PROP_COUNT], float st
+/*
+ * Sets the measure function on the cssNode, when the
+ * aspectRatio rectangle is set. This helps calculate
+ * the height of the node based on width of parent and
```
"...width of [the] parent..."

also can you adjust the line width of the comment to be 80 chars?

paramaggarwal replied:
fixed

ide commented on React/Views/RCTShadowView.m:
```diff
@@ -86,11 +86,46 @@ static void RCTProcessMetaProps(const float metaProps[META_PROP_COUNT], float st
+  result.dimensions[CSS_WIDTH] = width;
+  result.dimensions[CSS_HEIGHT] = computedHeight;
+  return result;
+}
```
Ideally, the aspect ratio would support views that have a known height (either hardcoded height from the stylesheet, or `alignSelf: stretch` in a horizontal parent) and unknown width. Any way to support this?

paramaggarwal replied:
Ah, good catch. Unfortunately this needs work on the `css-layout` side. Currently because this was intended for text layout, it only deals with height based on width scenario...

paramaggarwal commented:
Folks, please review! Also, if there is anything that should be done differently here, please share!

paramaggarwal commented:
Please review/share feedback!

hayeah commented:
I've been using aspectRatio from @paramaggarwal's branch, and have not run into any problem. Absolutely essential! :+1:

paramaggarwal commented:
Folks, please review!

hayeah commented:
ping @vjeux @ide @sahrens. Could aspectRatio make it into 0.10?

ide commented:
It won't go into 0.10 because we've cut the RC and this isn't an urgent bugfix. I don't think it will go into 0.11 either because that release will focus on stability, but it could make it into 0.12-rc.

For me to feel good about this commit I'd want aspectRatio to work in both dimensions. E.g. if you have an image with aspect ratio 4:3 and give it `height: 300`, it should be rendered at 400x300. Likewise it should be rendered at 400x300 if you specify `width: 400` too.

paramaggarwal commented:
@ide Good point! That needs a change in css-layout - will leave this open till we get around to supporting it there.

paramaggarwal commented:
Till then, could someone please pull #1776? It was imported but never showed up on `master`. With that we can use this component separately as a library. It's merged upstream also: https://github.com/facebook/css-layout/pull/84.

paramaggarwal commented:
Anybody would like to help out on https://github.com/facebook/css-layout/pull/118? The idea is to make `measure` more generic and be more like `sizeThatFits` from iOS. (ref: [`sizeThatFits`](https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIView_Class/#//apple_ref/occ/instm/UIView/layoutSubviews))

hayeah commented:
@paramaggarwal is there another issue I should watch? Would really like to see this happen...

paramaggarwal commented:
@hayeah Long story short, this PR depended on `measure` from css-layout to override layout when needed - for example aspect ratio. But `measure` was originally meant for only text layout, which means it only works to calculate height, given a width. I tried my hand at adding support to `measure` to also work the other way - calculate the width, given a height.

Basically the goal was to make `measure` a more generic way to override layout whenever a component needs it. Just like `sizeThatFits:` from [iOS UIView](https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIView_Class/#//apple_ref/occ/instm/UIView/sizeThatFits:). But I tried hard (https://github.com/facebook/css-layout/pull/118) and failed to make css-layout work that way. I don't have much knowledge of how css-layout works.

Currently I am making do with having a fork of RN with this particular commit https://github.com/facebook/react-native/pull/1776 merged in and then just having a small custom component called `RigidBox` to maintain aspect ratio of my images.

iostalk commented:
Very well done! But where is the doc?

wootwoot1234 commented:
Any update on this?  I could really use this in my project right now. :)

paramaggarwal commented:
@iostalk @wootwoot1234 The changes in this PR are able to support aspect ratio only in one dimension - for it to make it into core, it needs to support both vertical and horizontal dimensions. Unfortunately that needs serious rework on the `css-layout` repo.

alfrednerstu commented:
Also needing this badly, what are the latest developments? Is there a way I can get this to work @paramaggarwal ?

paramaggarwal commented:
@alfrednerstu I should probably publish what I have as an OSS component. Haven't gotten around to doing it...

paramaggarwal commented:
Hey guys, here is the generic component I have been using to make an arbitrary container who's height is always in a particular ratio to its width. The width is equal to the width of the parent. Great for use with images where you want them to stick to a particular aspect ratio.

Here is the code with the usage as a comment: https://gist.github.com/paramaggarwal/b9baa00fd7bee9a2948f

Feel free to use the code anywhere and also make it into an OSS component.

paramaggarwal commented:
The changed needed in `css-layout` are now in: https://github.com/facebook/css-layout/pull/154 - but I'll need to integrate (My colleague @obipawan has already done it internally) an Android version also in this PR. Will have to also add an example in UIExplorer.

Should I make the change only for `Image` component, or generically available on `View`?

polarathene commented:
@paramaggarwal  It would be nice as generic. I have a use case with react-native-video component where I'm wanting to use flex:1 for the width but maintain aspect ratio without having to provide a fixed height.

I'm curious if this feature might help towards implementing flex-grow/shrink features?

mschipperheyn commented:
This feels a bit like css's: background-size:cover. Ideally, you would be able to specify how to position the image in the defined window. Normally, a centered approach would be best
